### PR TITLE
SailBot simulator

### PR DESCRIPTION
**Pull request by Zimrahin**
# SailBot Simulator

This pull request introduces a sailboat simulator to the project. The simulator uses a state-space model with the following variables:

- *x*: x position in a Cartesian coordinate system
- *y*: y position in a Cartesian coordinate system
- *θ*: heading of the boat
- *v*: linear velocity of the boat
- *ω*: angular velocity of the boat

The simulator takes two inputs: the angle of the rudder and the opening of the mainsheet. It computes the actual angle of the sails based on the mainsheet opening and the apparent wind direction. The apparent wind direction is derived from the boat's linear velocity, heading, and the true wind velocity vector. 

**Modifications to the GUI:**
The JavaScript GUI used to control the actual SailBot has been adapted for the simulator, including features like:
- Interactive display of apparent wind direction
- Rudder angle and sail angle visualization

**Protocol and Firmware Adjustments:**
- Wind angle, rudder angle, and sail opening are communicated using packets of 2 bytes, 1 byte, and 1 byte, respectively, managed in `protocol.py`.
- The [SailBot firmware](https://github.com/DotBots/DotBot-firmware/tree/main/projects/03app_sailbot) requires modifications for compatibility with the updated serial protocol.

**Additional Feature:**
- Editing the rudder angle through string inputs is now possible, which may conflict with GUI operations on the actual SailBot. This conflict is due to the GUI being originally designed not to send commands when the rudder is at zero.

**Note**
For stability of the model, the time-step must be set to less than 0.04 seconds. By default, a time-step of 0.01 seconds is hard coded.

## Installation and Execution

To install and run the simulator:

1. Clone the repository.
2. Inside the main folder, run `pip install -e .`.
3. Navigate to `/dotbot/frontend` and run `npm install` followed by `npm run build`.
4. Open two terminals:
   - In one, navigate to `/dotbot/frontend` and run `npm start`.
   - In the other, in the main folder, run `dotbot-controller -p sailbot-simulator`.

## Visualization

Simulation Visualization

## References

- [Modeling, control and state estimation for an autonomous sailboat](https://www.researchgate.net/profile/Mohamed-Mourad-Lafifi/post/Is_it_possible_to_estimate_the_state_space_model_only_with_measurements_over_time/attachment/5e28a8e33843b093839ca513/AS%3A850233953103872%401579722979658/download/Modeling%2C+control+and+state-estimation+for+an+autonomous+sailboat.pdf)
- [Robotic Sailing 2013 Proceedings of the 6th International Robotic Sailing Conference](https://www.ensta-bretagne.fr/lebars/wrsc2013/0142.pdf)

### CODE REVIEW

**Comment by aabadie** [comment]
Can you rebase?

File: dotbot/protocol.py [hunk 1]
```diff
@@ -14,17 +14,17 @@
...
-    LH2_RAW_DATA = 2
-    LH2_LOCATION = 3
-    ADVERTISEMENT = 4
-    GPS_POSITION = 5
-    DOTBOT_DATA = 6
-    CONTROL_MODE = 7
-    LH2_WAYPOINTS = 8
-    GPS_WAYPOINTS = 9
-    SAILBOT_DATA = 10
+    CMD_MOVE_RAW    = 0
```
Comment (aabadie) [inline dotbot/protocol.py:17]: you don't need to align. That creates unrelated changes, making the review harder.

File: dotbot/protocol.py [hunk 2]
```diff
@@ -267,29 +267,38 @@ def from_bytes(bytes_) -> ProtocolData:
 
 @dataclass
 class SailBotData(ProtocolData):
-    """Dataclass that holds direction and GPS data and heading from SailBot application."""
+    """Dataclass that holds SailBot data from SailBot application."""
 
+    # Initialized as 0xFFFF because then at controller .py: -500 <= payload.values.direction <= 500
     direction: int = 0xFFFF
-    latitude: int = 0
+    latitude:  int = 0
```
Comment (aabadie) [inline dotbot/protocol.py:274]: unrelated change. I believe black will reformat that anyway

File: dotbot/sailbot_simulator.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+                for sailbot in self.sailbots:
+                    for byte in sailbot.update():
+                        self.callback(byte.to_bytes(length=1, byteorder="little"))
+
+                next_time = current_time + delta_t
+
+    def write(self, bytes_):
+        # write bytes on the fake serial, similar to the real gateway
+        for sailbot in self.sailbots:
+            sailbot.decode_serial_input(bytes_)
```
Comment (aabadie) [inline dotbot/sailbot_simulator.py:219]: There's a missing newline here but that should be catched by the CI already

File: dotbot/controller.py [hunk 4]
```diff
@@ -95,28 +96,30 @@ class Controller:
...
+            #     last_seen=time.time(),
+            # ),
+            # "0000000000000004": DotBotModel(
+            #     address="0000000000000004",
+            #     application=ApplicationType.SailBot,
+            #     last_seen=time.time(),
+            #     wind_angle=135,
+            #     rotation=49,
+            #     gps_position=DotBotGPSPosition(latitude=48.832313766146896, longitude=2.4126897594949184),
+            # ),
```
Comment (aabadie) [inline dotbot/controller.py:122]: This is mainly unrelated changes, except the wind_angle and rotation of the last one. Can you keep is as it was before but just add the 2 new lines

File: dotbot/controller.py [hunk 5]
```diff
@@ -147,25 +150,28 @@ def on_byte_received(byte):
...
-            if int.from_bytes(byte, byteorder="little") != PROTOCOL_VERSION:
-                raise SerialInterfaceException("Handshake failed")
-
-        self.serial = SerialInterface(
-            self.settings.port, self.settings.baudrate, on_byte_received
-        )
+        if self.settings.port == "sailbot-simulator":
+            self.serial = SailBotSimSerialInterface(on_byte_received)
+        else:
+            async def _wait_for_handshake(queue):
```
Comment (aabadie) [inline dotbot/controller.py:156]: in #120, IIRC, all this indent out of large blocks is not required. The handshake can also be handled by the simulator.

**Review by aabadie — COMMENTED** [review]
A first round of reviews

**Comment by malishav** [comment]
Just pushed the autonomous mode on top of this branch done by @Zimrahin , so the PR contains both manual mode and autonomous mode.

**Review by aabadie — APPROVED** [review]
